### DISCUSSION TITLE
Building a SPARQL Endpoint

### DISCUSSION THREAD

**Discussion by miroslavliska** [eclipse-rdf4j/rdf4j#3776] [Q&A]
Dear sirs, 

I am trying to create a SPARQL Endpoint on top of our RDF database and I am using rdf4j (It is the first knowledge graph for our (Slovakian) government). https://github.com/datova-kancelaria/znalosti.gov.sk

What I don't know to implement is to return TupleQueryResult in various formats, such as TupleQueryResultFormat supports. However, I haven't found any examples fit for this situation. 

```
@GetMapping("/api/sparql")
public List<BindingSet> execute(String query) throws IOException, UnsupportedRDFormatException, FileNotFoundException, ParserConfigurationException, TransformerException {	
		List<BindingSet> resultList;
		query = "select * where {?s <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?o}";		
		try (RepositoryConnection conn = repository.getConnection()) {			
			TupleQuery tupleQuery = conn.prepareTupleQuery(query);			
			try (TupleQueryResult result = tupleQuery.evaluate()) {
			   resultList = QueryResults.asList(result);
			}							
			 return resultList;
}
```

However, this only returns  BindgindNames and not the values
![image](https://user-images.githubusercontent.com/37801000/161704365-8c63714d-1c5e-4ec9-a2a7-d372ecfb8d1d.png)

May I please ask you, how to use this code together with the TupleQueryResultFormat, in order to get CSV output for example?

I found for instance these examples
https://www.programcreek.com/java-api-examples/?api=org.eclipse.rdf4j.query.resultio.TupleQueryResultFormat
but none fits my code -  I mean to simply return selected format with TupleQueryResultFormat.

thank you all for the guidance

**Comment by JervenBolleman** [comment]
Hi @miroslavliska you might want to have a look at the spring-boot component that is in 4.0.0-M2 release.
[spring-boot-sparql-web](https://github.com/eclipse/rdf4j/tree/develop/spring-components/spring-boot-sparql-web)

Maybe you could test it out for us?

**Reply by miroslavliska** [reply]
Thank you very much.  It seems to be very interesting. I will try to run it, but it will take me a while.

**Reply by miroslavliska** [reply]
Hi @JervenBolleman. I have added the QueryResponder class to my project. After the successful run, I am only able to get a TRIX format with a select query. I am trying to get JSONLD or RDF/XML, but I a get 

```
There was an unexpected error (type=Internal Server Error, status=500).
Did not recognise RDF format object RDF/XML (mimeTypes=application/rdf+xml, application/xml, text/xml; ext=rdf, rdfs, owl, xml)
org.eclipse.rdf4j.rio.UnsupportedRDFormatException: Did not recognise RDF format object RDF/XML (mimeTypes=application/rdf+xml, application/xml, text/xml; ext=rdf, rdfs, owl, xml)
	at org.eclipse.rdf4j.rio.Rio.lambda$unsupportedFormat$0(Rio.java:690)
```

My pom.xml project contains follows RDF4J dependencies
```
   <properties>
        <java.version>11</java.version>
        <vaadin.version>14.4.6</vaadin.version>
        <rdf4j.version>4.0.0-M2</rdf4j.version>
    </properties>

 <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>com.vaadin</groupId>
                <artifactId>vaadin-bom</artifactId>
                <version>${vaadin.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
         
         <dependency>
            <groupId>org.eclipse.rdf4j</groupId>
            <artifactId>rdf4j-spring-components</artifactId>
            <version>${rdf4j.version}</version>
        </dependency>

       <dependency>
            <groupId>org.eclipse.rdf4j</groupId>
            <artifactId>spring-boot-sparql-web</artifactId>
            <version>${rdf4j.version}</version>
        </dependency>
 
        </dependencies>
       
    </dependencyManagement>

    <dependencies>
       <dependency>
            <groupId>org.eclipse.rdf4j</groupId>
            <artifactId>rdf4j-sail-memory</artifactId>
         <!--   <version>3.2.3</version> -->
            <version>${rdf4j.version}</version>
        </dependency>

        <dependency>
            <groupId>org.eclipse.rdf4j</groupId>
            <artifactId>rdf4j-repository-manager</artifactId>
            <version>${rdf4j.version}</version>
        </dependency>

        <dependency>
            <groupId>org.eclipse.rdf4j</groupId>
            <artifactId>rdf4j-repository-sail</artifactId>
            <version>${rdf4j.version}</version>
        </dependency>

        <dependency>
            <groupId>org.eclipse.rdf4j</groupId>
            <artifactId>rdf4j-repository-http</artifactId>
            <version>${rdf4j.version}</version>
        </dependency>
        
        <dependency>
            <groupId>org.eclipse.rdf4j</groupId>
            <artifactId>rdf4j-rio-api</artifactId>
            <version>${rdf4j.version}</version>
        </dependency>
        
        <dependency>
            <groupId>org.eclipse.rdf4j</groupId>
            <artifactId>rdf4j-queryresultio-api</artifactId>
            <version>${rdf4j.version}</version>
        </dependency>  

</dependencies>

```

Can you give me please some advice please on what to check?

**Reply by miroslavliska** [reply]
OK, I have made small progress. Now I can run also SPARQL construct queries and get RDF/XML. However I still don’t have control over the requests, so I  will keep debugging. 
![image](https://user-images.githubusercontent.com/37801000/165826983-ee0407c6-1b1a-4cad-a1e3-b9a5ce6d66fe.png)

**Reply by JervenBolleman** [reply]
Sorry, I missed your earlier message. Did you fix it by adding more rio format specific dependencies?

**Reply by miroslavliska** [reply]
Yes. I added the rdf4j-rio-rdfxml dependency to my pom.xml (as well as other formats). Once I will finish our endpoint I will share the link. However, the rdf4j code for a SPARQL endpoint is perfect! Here is the working jsonld output .
![image](https://user-images.githubusercontent.com/37801000/165832660-fdbb5104-f7ed-489b-bdca-8d5386f817e1.png)

**Comment by naturzukunft** [comment]
@JervenBolleman, some day i have to compare my solution with your newest one. I was using a very early version of yours and apopted it for my problem. Currently it's in rdf-pub as an apache-camel endpoint: 
[RouteQueryCollection](https://gitlab.com/linkedopenactors/rdf-pub/-/blob/develop/rdf-pub-core/src/main/java/org/linkedopenactors/rdfpub/usecases/querycollection/RouteQueryCollection.java) & [ContentEnricherSparql](https://gitlab.com/linkedopenactors/rdf-pub/-/blob/develop/rdf-pub-core/src/main/java/org/linkedopenactors/rdfpub/usecases/querycollection/ContentEnricherSparql.java).

**Reply by naturzukunft** [reply]
i was excited to read about rdf4j-spring in rdf4j 4.0.0 today. hopefully i will have time to play with it soon ;-)
